// nuxt-jsonld/dist/module.js
// ../../node_modules/tsup/assets/esm_shims.js
import { fileURLToPath } from "url";
import path from "path";
var getFilename = () => fileURLToPath(import.meta.url);
var getDirname = () => path.dirname(getFilename());
var __dirname = /* @__PURE__ */ getDirname();

// src/module.ts
import { resolve } from "pathe";
import { defineNuxtModule, addPlugin, addImports } from "@nuxt/kit";
var module_default = defineNuxtModule({
  meta: {
    name: "nuxt-jsonld",
    compatibility: {
      nuxt: "^3.0.0-rc.8",
      bridge: false
    }
  },
  setup(options, nuxt) {
    const runtimeDir = resolve(__dirname, "./runtime");
    const composable = resolve(__dirname, "./runtime/composable");
    nuxt.options.build.transpile.push(runtimeDir);
    nuxt.options.alias["#jsonld"] = composable;
    addImports([{ name: "useJsonld", as: "useJsonld", from: composable }]);
    if (!options.disableOptionsAPI) {
      addPlugin(resolve(runtimeDir, "plugin"));
    }
  }
});
export {
  module_default as default
};
